contributors/design-proposals/cpu-affinity-numa-awareness.md

```diff
@@ -0,0 +1,216 @@
+	// associated with this node.
+	// Example: 0-3 or 0,2,4,6
+	// The values are expressed in the List Format syntax specified
+	// here: http://man7.org/linux/man-pages/man7/cpuset.7.html
+	CPUSet string
+}
+```
+
+## Node Configuration
+
+### Isolating host processes
+
+By default, load balancing is done across all CPUs, except those marked isolated
+using the kernel boot time `isolcpus=` argument.  When configuring a node to support
+CPU and NUMA affinity, many operators may wish to isolate host processes to particular
+cores.
+
+It is recommended that operators set a CPU value for `--system-reserved`
+in whole cores that aligns with the set of cpus that are made available to the default
+kernel scheduling algorithm.  If an operator is on a `systemd` managed platform, they
+may choose instead to set the `CPUAffinity` value for the root slice to the set of CPU
+cores that are reserved for the host processes.
+
+**TODO**
+
+1. how should `kubelet` discover the reserved `cpu-set` value?
+1. in a numa system, `kubelet` reservation for memory needs to be removed from
+a particular numa node capacity so numa node allocatable is as expected. 
+
+### Configuring Taints
+
+The following `Taint` keys are defined to enable CPU pinning and NUMA awareness.
+
+#### CPUAffinity
+
+* Effect: `NoScheduleNoAdmitNoExecute`
+* Potential values:
+ * `dedicated`
+
+If `dedicated`, all pods that match this taint will require dedicated compute resources. Each
+pod bound to this node must request CPU in whole cores.  The CPU limit must equal the request.
+
+#### NUMACPUAffinity
+
+* Effect: `NoScheduleNoAdmitNoExecute`
+* Potential values:
+ * `strict`
+
+If `strict`, all pods that match this taint must request CPU (whole or fractional cores) that
+fit a single NUMA node `cpu` allocatable.
+
+#### NUMAMemoryPolicy
+
+* Effect: `NoScheduleNoAdmitNoExecute`
+* Potential values:
+ * `strict`
+ * `preferred`
+
+If `strict`, all pods that match this taint must request `memory` that fits it's assigned
+NUMA node `memory` allocatable.
+
+If `preferred`, all pods that match this taint are not required to have their `memory` request
+fit it's assigned NUMA node `memory` allocatable.
+
+## Pod Specification
+
+### API changes
+
+The following API changes are proposed to the `PodSpec` to allow CPU and NUMA affinity to be defined.
+
+```
+// PodSpec is a description of a pod
+type PodSpec struct {
+...
+	// NodeName is a request to schedule this pod onto a specific node.  If it is non-empty,
+	// the scheduler simply schedules this pod onto that node, assuming that it fits resource
+	// requirements.
+	// +optional
+	NodeName string
+	// Identifies a NUMA node that affines the pod.  If it is non-empty, the value must
+	// correspond to a particular NUMA node on the same node that the pod is scheduled against.
+	// This value is only set if either the `CPUAffinity` or `NUMACPUAffinity` tolerations
+	// are present on the pod.
+	// +optional
+	NUMANodeID string
```

Comment:
I've said before and I still assert - this is a terrible plan.  Offering up this much control is going to kill off any chances we have of actually doing the right thing automatically for most users.  This is an attractive nuisance (https://en.wikipedia.org/wiki/Attractive_nuisance_doctrine) - people are going to get hurt by it.  If the borg experience counts for anything, the lesson is that saying "yes" to things like this turned out terribly, and is a drag on the system.

I proposed a less first-class alternative.  Something like: annotate pods and run a highly-privileged agent on the node react to pods being scheduled by programming cpusets and calling numactl and whatever else it needs to do.  Make this VERY explicit to use and DO NOT make it part of the API.

If we need to collect NUMA status and use something about NUMA as a hint to scheduling, I can MAYBE see this, maybe even a different scheduler enitrely.

If we need to sell "dedicated caches", then we should add that as resource.  If I need NUMA because I want an LLC to myself, let me spec that.

If we need to sell "link to high-perf IO device", then we should add that as resource.

I'm trying to not overreact, but I have a visceral response to this.